nio/client/async_client.py

```diff
@@ -2291,6 +2300,26 @@ async def upload(
 
             filesize (int, optional): Size in bytes for the file to transfer.
                 If left as ``None``, some servers might refuse the upload.
+
+        It's common to use this alongside :py:meth:`room_send`. An example of
+        uploading a plain text file follows, but the principle is the same for
+        media, you just need to add an additional "info" key to the content.
+        See `the Matrix client-server spec <https://matrix.org/docs/spec/client_server/r0.6.0#m-room-message-msgtypes>`_
+        for more details.
+
+        Example:
+            async with aiofiles.open("sample.py", "r+b") as f:
+                resp, err = await client.upload(f, content_type="text/plain", filename="hello.py")
```

Comment:
It doesn't return an response and error, it returns a response which can be an error or success response and the encryption keys that were used to encrypt the file if encrypt was set to true.

We might also want to add the `file_size` argument, seems like the last release of aiofiles added support for this. If the argument isn't provided a bare Synapse throws errors.